src/Files.App/Helpers/AdaptiveLayoutHelpers.cs

```diff
@@ -3,124 +3,128 @@
-				}
-				else
-				{
-					layoutDetails();
-				}
-
-				return true;
 			}
+		}
+
+		private static Layouts GetAdaptiveLayout(string path, IList<ListedItem> filesAndFolders)
+		{
+			var pathLayout = GetPathLayout(path);
+			if (pathLayout is not Layouts.None)
+				return pathLayout;
+
+			return GetContentLayout(filesAndFolders);
+		}
+
+		private static Layouts GetPathLayout(string path)
+		{
+			var iniPath = IO.Path.Combine(path, "desktop.ini");
+
+			var iniContents = NativeFileOperationsHelper.ReadStringFromFile(iniPath)?.Trim();
+			if (string.IsNullOrEmpty(iniContents))
+				return Layouts.None;
+
+			var parser = new IniParser.Parser.IniDataParser();
+			parser.Configuration.ThrowExceptionsOnError = false;
+			var data = parser.Parse(iniContents);
+			if (data is null)
+				return Layouts.None;
+
+			var viewModeSection = data.Sections.FirstOrDefault(IsViewState);
+			if (viewModeSection is null)
+				return Layouts.None;
 
-			return false;
+			var folderTypeKey = viewModeSection.Keys.FirstOrDefault(IsFolderType);
+			if (folderTypeKey is null)
+				return Layouts.None;
+
+			return folderTypeKey.Value switch
+			{
+				"Pictures" => Layouts.Grid,
+				"Videos" => Layouts.Grid,
+				_ => Layouts.Detail,
+			};
+
+			static bool IsViewState(SectionData data)
+				=> "ViewState".Equals(data.SectionName, StringComparison.OrdinalIgnoreCase);
+
+			static bool IsFolderType(KeyData data)
+				=> "FolderType".Equals(data.KeyName, StringComparison.OrdinalIgnoreCase);
+		}
+
+		private static Layouts GetContentLayout(IList<ListedItem> filesAndFolders)
+		{
+			int itemCount = filesAndFolders.Count;
+			if (filesAndFolders.Count is 0)
+				return Layouts.None;
+
+			float folderPercentage = 100f * filesAndFolders.Count(IsFolder) / itemCount;
+			float imagePercentage = 100f * filesAndFolders.Count(IsImage) / itemCount;
+			float mediaPercentage = 100f * filesAndFolders.Count(IsMedia) / itemCount;
+			float miscPercentage = 100f - (folderPercentage + imagePercentage + mediaPercentage);
+
+			if (folderPercentage + miscPercentage > LargeThreshold)
+				return Layouts.Detail;
+			if (imagePercentage > ExtraLargeThreshold)
+				return Layouts.Grid;
+			if (imagePercentage <= MediumThreshold)
+				return Layouts.Detail;
+			if (100f - imagePercentage <= SmallThreshold)
+				return Layouts.Detail;
+			if (folderPercentage + miscPercentage <= ExtraSmallThreshold)
+				return Layouts.Detail;
+			return Layouts.Grid;
+
+			static bool IsFolder(ListedItem item)
+				=> item.PrimaryItemAttribute is StorageItemTypes.Folder;
+
+			static bool IsImage(ListedItem item)
+				=> !string.IsNullOrEmpty(item.FileExtension)
+				&& ImagePreviewViewModel.ContainsExtension(item.FileExtension.ToLowerInvariant());
+
+			static bool IsMedia(ListedItem item)
+				=> !string.IsNullOrEmpty(item.FileExtension)
+				&& MediaPreviewViewModel.ContainsExtension(item.FileExtension.ToLowerInvariant());
+		}
+
+		private enum Layouts
+		{
+			None,
+			Detail,
+			Grid,
 		}
```

Comment:
Is there no Layouts type enum somewhere else? If not, does it really belong here and not inside a specialized enum stored inside its own file?

Reply:
This enum is very specific because it contains the value None. This value is used to check the next case or do nothing otherwise. This is a class implementation trick that should have no impact on the rest of the code.

Reply:
A comment about this trick may be useful here, so it's not replicated anywhere else by accident if the knowledge get lost.

Reply:
A comment have been added.